pull/pull_requests.go

```diff
@@ -23,57 +23,125 @@ import (
+	List(ctx context.Context, owner, repo string, opts *github.PullRequestListOptions) ([]*github.PullRequest, *github.Response, error)
+}
 
+// GetOpenPullRequestsForSHA returns all open pull requests where the HEAD of the source branch
+// matches the given SHA.
+func GetOpenPullRequestsForSHA(ctx context.Context, client GitHubClient, owner, repo, sha string) ([]*github.PullRequest, error) {
+	logger := zerolog.Ctx(ctx)
 	var results []*github.PullRequest
-	for _, pr := range prs {
-		if pr.GetState() == "open" && pr.GetHead().GetSHA() == SHA {
-			results = append(results, pr)
+	opts := &github.ListOptions{PerPage: 100}
+
+	for {
+		prs, resp, err := client.ListPullRequestsWithCommit(ctx, owner, repo, sha, opts)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list pull requests for repository %s/%s", owner, repo)
+		}
+
+		for _, pr := range prs {
+			if pr.GetState() == "open" && pr.GetHead().GetSHA() == sha {
+				logger.Debug().Msgf("found open pull request with sha %s", pr.GetHead().GetSHA())
+				results = append(results, pr)
+			}
 		}
+
+		if resp.NextPage == 0 {
+			break
+		}
+		opts.Page = resp.NextPage
 	}
+
 	return results, nil
 }
 
-func ListOpenPullRequestsForRef(ctx context.Context, client *github.Client, owner, repoName, ref string) ([]*github.PullRequest, error) {
+// ListOpenPullRequestsForSHA returns all open pull requests where the HEAD of the source branch
+// matches the given SHA by fetching all open PRs and filtering.
+func ListOpenPullRequestsForSHA(ctx context.Context, client GitHubClient, owner, repo, sha string) ([]*github.PullRequest, error) {
+	logger := zerolog.Ctx(ctx)
 	var results []*github.PullRequest
+	opts := &github.PullRequestListOptions{
+		State:       "open",
+		ListOptions: github.ListOptions{PerPage: 100},
+	}
+
+	for {
+		prs, resp, err := client.List(ctx, owner, repo, opts)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list pull requests for repository %s/%s", owner, repo)
+		}
+
+		for _, pr := range prs {
+			if pr.Head.GetSHA() == sha {
+				logger.Debug().Msgf("found open pull request with sha %s", pr.Head.GetSHA())
+				results = append(results, pr)
+			}
+		}
+
+		if resp.NextPage == 0 {
+			break
+		}
+		opts.Page = resp.NextPage
+	}
+
+	return results, nil
+}
+
+// GetAllPossibleOpenPullRequestsForSHA attempts to find all open pull requests
+// associated with the given SHA using multiple methods in case we are dealing with a fork
+func GetAllPossibleOpenPullRequestsForSHA(ctx context.Context, client GitHubClient, owner, repo, sha string) ([]*github.PullRequest, error) {
 	logger := zerolog.Ctx(ctx)
 
-	ref = strings.TrimPrefix(ref, "refs/heads/")
+	prs, err := GetOpenPullRequestsForSHA(ctx, client, owner, repo, sha)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get open pull requests matching the SHA")
+	}
 
+	if len(prs) == 0 {
+		logger.Debug().Msg("No pull requests associated with the check run, searching by SHA")
```

Comment:
nit: Let's not mention check runs here specifically, since this could be used for other things too